File: java/S1_ConnectSensor/src/TemperatureSensor.java
/*
 *                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This code is part of example scenario 1 'Connect a Sensor' of the
 * ADLINK Edge SDK. For a description of this scenario see the
 * 'Edge SDK User Guide' in the /doc directory of the Edge SDK instalation.
 *
 * For instructions on building and running the example see the README
 * file in the Edge SDK installation directory.
 */

import java.io.Closeable;
import java.util.Random;

import com.adlinktech.datariver.DataRiver;
import com.adlinktech.datariver.IotNvp;
import com.adlinktech.datariver.IotNvpSeq;
import com.adlinktech.datariver.IotValue;
import com.adlinktech.datariver.JSonTagGroupRegistry;
import com.adlinktech.datariver.JSonThingClassRegistry;
import com.adlinktech.datariver.JSonThingProperties;
import com.adlinktech.datariver.Thing;

public class TemperatureSensor implements Closeable {
	private final DataRiver dr;
	private final Thing thing;
	private final String thingPropertiesURI;
	static final int SAMPLE_DELAY_MS = 100;

	public TemperatureSensor(final String thingPropertiesURI) {
        System.out.println("Temperature Sensor started");
		this.thingPropertiesURI = thingPropertiesURI;
		this.dr = DataRiver.getInstance();
		this.thing = createThing();
	}

	private String getAbsFileURI(final String fileName) {
	    final String dir = System.getProperty("user.dir");
	    final String prefix = "file://";
		return prefix.concat(dir).concat("/").concat(fileName);
	}

	private void cleanup() {
		if (dr != null) {
			dr.close();
		}
		System.out.println("Temperature Sensor stopped");
	}

	private Thing createThing() {
	    final JSonTagGroupRegistry tgr = new JSonTagGroupRegistry();
		tgr.registerTagGroupsFromUri(getAbsFileURI("definitions/TagGroup/com.adlinktech.example/TemperatureTagGroup.json"));
		this.dr.addTagGroupRegistry(tgr);

		final JSonThingClassRegistry tcr = new JSonThingClassRegistry();
		tcr.registerThingClassesFromUri(getAbsFileURI("definitions/ThingClass/com.adlinktech.example/TemperatureSensorThingClass.json"));
		dr.addThingClassRegistry(tcr);

		final JSonThingProperties tp = new JSonThingProperties();
		tp.readPropertiesFromUri(thingPropertiesURI);
		
		return dr.createThing(tp);
	}

	private void writeSample(final float temperature) {
	    final IotValue tempV = new IotValue();
		tempV.setFloat32(temperature);

		final IotNvpSeq sensorData = new IotNvpSeq();
		sensorData.add(new IotNvp("temperature", tempV));

		thing.write("temperature", sensorData);
	}

	public void run(final long runningTime) {
	    final Random random = new Random();
		double sampleCount = (runningTime * 1000.0) / SAMPLE_DELAY_MS;
		float actualTemp = 21.5f;

		while (sampleCount-- > 0) {
			actualTemp += (random.nextInt(10) - 5)/ 5.0;
			writeSample(actualTemp);
			try {
                Thread.sleep(SAMPLE_DELAY_MS);
            } catch (InterruptedException e) {
                System.out.println("TemperatureSensor is interrupted. Exiting");
                return;
            }
		}
	}

	@Override
	public void close() {
		cleanup();
	}

}

File: nodejs/S1_ConnectSensorJsontgc/lib/temperature_sensor.js
/*
 *                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2020 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This code is part of example scenario 1 'Connect a Sensor' of the
 * ADLINK Edge SDK. For a description of this scenario see the
 * 'Edge SDK User Guide' in the /doc directory of the Edge SDK instalation.
 *
 * For instructions on running the example see the README
 * file in the Edge SDK installation directory.
 */

'use strict';

const datariver = require('@adlinktech/datariver');
const tg = require('./TemperatureTagGroup');
  
const util = require('util');
const sleep = util.promisify(setTimeout);

const SAMPLE_DELAY_MS = 100;

function getRandomInt(max) {
    return Math.floor(Math.random() * Math.floor(max));
}

class TemperatureSensor {
    
    // Initializing
    constructor(thingPropertiesUri) {
        this._thingPropertiesUri = thingPropertiesUri;
        
        this._datariver = null;
        this._thing = null;
    }

    // Enter the runtime context related to the object
    attachToDataRiver() {
        this._datariver = datariver.DataRiver.getInstance();
        this._thing = this.createThing();
        
        console.log('Temperature Sensor started');
    }

    // Exit the runtime context related to the object
    detachFromDataRiver() {
        if (this._datariver !== null) {
            this._datariver.close();
        }
        console.log('Temperature Sensor stopped');
    }

    createThing() {
        // Create and Populate the TagGroup registry with JSON resource files.
        const tgr = new datariver.JSonTagGroupRegistry();
        tgr.registerTagGroupsFromUri('file://definitions/TagGroup/com.adlinktech.example/TemperatureTagGroup.json');
        this._datariver.addTagGroupRegistry(tgr) ;

        // Create and Populate the ThingClass registry with JSON resource files.
        const tcr = new datariver.JSonThingClassRegistry();
        tcr.registerThingClassesFromUri('file://definitions/ThingClass/com.adlinktech.example/TemperatureSensorThingClass.json');
        this._datariver.addThingClassRegistry(tcr);

        // Create a Thing based on properties specified in a JSON resource file.
        const tp = new datariver.JSonThingProperties();
        tp.readPropertiesFromUri(this._thingPropertiesUri);
        
        const thing = this._datariver.createThing(tp);
        thing.getOutputHandler('temperature').validateSamples = true;
        return thing;
    }

    async writeSample(temperature) {
        try {
            const sensorData = new tg.Temperature();

            sensorData.temperature = temperature;

            // Write the data
            this._thing.write('temperature', tg.Temperature.to_nvp_seq(sensorData));

        } catch(error) {
            console.log('Error writing data: %O', error);
        }
    }
    
    async run(runningTime) {
        try {
            let sampleCount = (runningTime * 1000.0) / SAMPLE_DELAY_MS;
            let actualTemp = 21.5;

            while(sampleCount-- > 0) {
                actualTemp += (getRandomInt(10) - 5)/ 5.0;
                await this.writeSample(actualTemp);

                // Give middleware some time to finish writing samples
                await sleep(SAMPLE_DELAY_MS);
            }

        } catch(error) {
            console.log('TemperatureSensor run error: %O', error);
        }
    }
}
    
function getCommandLineParameters() {
    const argv = require('yargs')
        .usage('Usage: $0 [options]')
        .alias('t','thing-properties-uri')
        .nargs('t',1)
        .default('t', 'file://./config/TemperatureSensorProperties.json')
        .describe('t', 'Thing properties URI')

        .alias('r', 'running-time')
        .nargs('r', 1)
        .default('r', 60)
        .describe('r', 'Running time (seconds, 0 = infinite)')

        .help('h')
        .alias('h', 'help')
        .argv;

    return {
        thingPropertiesUri: argv['thing-properties-uri'],
        runningTime: argv['running-time']
    };
}

async function main() {
    // Get command line parameters
    const { thingPropertiesUri, runningTime } = getCommandLineParameters();

    const tempSensor = new TemperatureSensor(thingPropertiesUri);
    try {
        tempSensor.attachToDataRiver();
        await tempSensor.run(runningTime);
    } catch(error) {
        console.log('TemperatureSensor error: %O', error);
    } finally {
        tempSensor.detachFromDataRiver();
    }
}

main().then(() => console.log('TemperatureSensor done.')).catch(error => console.log('TemperatureSensor error: %O', error));